src/echo-common/src/poe-zone-tracker-service.ts

```diff
@@ -0,0 +1,465 @@
+          '|' +
+          this.currentZoneDelta.poeZoneInstance.areaTag +
+          '|' +
+          this.currentZoneDelta.poeZoneInstance.areaLevel.toString() +
+          '|' +
+          this.currentZoneDelta.poeZoneInstance.zoneSeed.toString()
+
+        if (
+          event.areaTag.toLowerCase().includes('hideout') ||
+          event.areaTag.toLowerCase().includes('town')
+        ) {
+          this.currentZoneDelta.poeZoneInstance.isZoneHideoutOrTown = true
+          zoneHash = zoneHash + '|' + event.time.getTime().toString()
+        } else {
+          this.currentZoneDelta.poeZoneInstance.isZoneHideoutOrTown = false
+        }
+
+        this.currentZoneDelta.poeZoneInstance.zoneHash = zoneHash
+
+        const existingZone = this.zones.get(zoneHash)
+        if (existingZone) {
+          this.currentZoneDelta.poeZoneInstance = existingZone
+          //TODO IF EXISTING ZONE IS FOUND, DO SOME LOGIC BASED ON DATE TO DETERMINE IF IT IS CLOSE ENOUGH
+          existingZone.zoneDeltas.push(this.currentZoneDelta)
+          this.zones.set(zoneHash, existingZone)
+        } else {
+          this.currentZoneDelta.poeZoneInstance.zoneDeltas.push(this.currentZoneDelta)
+          this.zones.set(zoneHash, this.currentZoneDelta.poeZoneInstance)
+        }
+      })
+  }
+
+  private zoneEntranceEventSubscribe() {
+    this.logEvents$
+      .pipe(
+        filter((e) => e.type == 'ZoneEntranceEvent'),
+        map((e) => e as PoeZoneEntranceEvent)
+      )
+      .subscribe((event) => {
+        if (this.currentZoneDelta === undefined) {
+          //Emit as event
+          throw new Error(
+            'Something went wrong. currentZoneDelta not instansiated correctly in ZoneEntranceEvent'
+          )
+        }
+
+        if (!this.currentZoneDelta.startLoadTime) {
+          //Emit as event
+          throw new Error(
+            'Something went wrong. currentZoneDelta.startLoadTime not set correctly in ZoneEntranceEvent'
+          )
+        }
+
+        this.currentZoneDelta.poeZoneInstance.zoneName = event.location
+        this.currentZoneDelta.enterTime = event.time
+        this.currentZoneDelta.enterSystemTime = event.systemUptime
+
+        const loadStart = this.currentZoneDelta.startLoadTime
+        this.currentZoneDelta.doneLoadTime = event.systemUptime
+        this.currentZoneDelta.deltaLoadTime = event.systemUptime - loadStart
+
+        if (this.characterService != undefined) {
+          const charName = 'BlaesendeBruno' //TODO from settings
+          this.characterSubscription = this.characterService
+            .character(charName) //TODO Find a solution to call this function to get non-stale data, probably more intelligent rate-limiter
+            .pipe(
+              filter((e) => e.type == 'result'), //MAKE SURE ERRORS ARE HANDLED ASWELL. PROBABLY USE OF + ZIP TO CHECK CURRENT ZONE IS STILL THE SAME IN CASE OF RATELIMIT
+              map((e) => e as SmartCacheResultEvent<PoeCharacter>),
```

Comment:
Is this really the way to get typescript to recognize the type, or am i missing something?

Reply:
194:196 can be replaced with `validResults` from https://github.com/PoeStack/poestack-sage/blob/main/src/echo-common/src/smart-cache-hooks.ts#L25

Reply:
`validResults` has a more strict type definition so typescript will recognize the types.